File: dcp-solutions/2024-11-16/solution.go
```markdown
type Node struct {
    Key   int
    Value interface{}
    Count int
    Prev  *Node
    Next  *Node
}

type LFUCache struct {
    Capacity int
    KeyToNode map[int]*Node
    CountToNodes map[int]*Node
    MinCount  int
}

func NewLFUCache(capacity int) *LFUCache {
    lfucache := &LFUCache{
        Capacity: capacity,
        KeyToNode: make(map[int]*Node),
        CountToNodes: make(map[int]*Node),
        MinCount:    1,
    }
    return lfucache
}

func (lfucache *LFUCache) Get(key int) interface{} {
    node, ok := lfucache.KeyToNode[key]
    if !ok {
        return nil
    }

    node.Count--
    if node.Count == 0 {
        delete(lfucache.CountToNodes, node.Count)
        lfucache.MinCount = lfucache.MinCount + 1
        for node.Next != nil && lfucache.MinCount == node.Count {
            prev := node.Prev
            next := node.Next

            prev.Next = next
            if next != nil {
                next.Prev = prev
            } else {
                lfucache.CountToNodes[lfucache.MinCount] = prev
            }

            delete(lfucache.KeyToNode, next.Key)
            delete(lfucache.CountToNodes, node.Count)

            node.Key = next.Key
            node.Value = next.Value
            node.Next = nil

        }
    }
    lfucache.CountToNodes[node.Count] = &node
    return node.Value
}

func (lfucache *LFUCache) Set(key int, value interface{}) {
    if _, ok := lfucache.KeyToNode[key]; ok {
        lfucache.Get(key)
    }

    if len(lfucache.KeyToNode) == lfucache.Capacity {
        minCount := lfucache.MinCount
        delete(lfucache.CountToNodes, minCount)

        node := lfucache.CountToNodes[minCount]
        for node.Next != nil {
            next := node.Next

            prev := node.Prev
            prev.Next = next
            if next != nil {
                next.Prev = prev
            } else {
                lfucache.CountToNodes[lfucache.MinCount] = prev
            }

            delete(lfucache.KeyToNode, next.Key)
        }
    }

    newNode := &Node{
        Key:   key,
        Value: value,
        Count: 1,
    }

    if lfucache.MinCount == 0 {
        lfucache.MinCount = 1
        lfucache.CountToNodes[1] = newNode
    } else {
        lfucache.CountToNodes[1] = newNode
    }

    lfucache.KeyToNode[key] = newNode

    newNode.Next = lfucache.CountToNodes[lfucache.MinCount]
    if lfucache.CountToNodes[lfucache.MinCount] != nil {
        lfucache.CountToNodes[lfucache.MinCount].Prev = newNode
    }
}
```

File: dcp-solutions/2024-11-16/solution.c
```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_KEY_LENGTH 128

typedef struct Node {
    char key[MAX_KEY_LENGTH];
    int value;
    int freq;
    struct Node* prev, * next;
} Node;

typedef struct LFUCache {
    Node* head;
    Node* tail;
    int size;
    int capacity;
    int frequencies[1000000];
} LFUCache;

LFUCache* new_LFUCache(int capacity) {
    LFUCache* lfu = (LFUCache*)malloc(sizeof(LFUCache));
    if(lfu == NULL) return NULL;
    lfu->capacity = capacity;
    lfu->head = NULL;
    lfu->tail = NULL;
    lfu->size = 0;
    memset(lfu->freqrequencies, 0, sizeof(int)*capacity);
    return lfu;
}

void removeNode(Node* node) {
    if(node == NULL || node->prev == NULL || node->next == NULL)
        return;

    node->prev->next = node->next;
    node->next->prev = node->prev;
}

void addToTail(Node* node) {
    if(lfu->tail == NULL)
        lfu->head = node;
    else
        lfu->tail->next = node;
    lfu->tail = node;
}

int getCount(int freq) {
    return lfu->freqrequencies[freq];
}

void updateFreq(int key, int value) {
    if(getCount(0) == 1 && lfu->size == 1)
        return;

    Node* currNode = lfu->head;
    while(currNode != NULL && strcmp(currNode->key, key) != 0){
        currNode = currNode->next;
    }
    
    if(currNode == NULL || lfu->tail == NULL ||strcmp(currNode->key, key) != 0) return;

    int freq = (currNode->freq + 1)%lfu->capacity;
    removeNode(currNode);
    if(getCount(freq) > getCount(0))
        freq++;
    updateFreq(key, value);
    lfu->size--;
    addToTail(lfu->head);
}

void set(int key, int value) {
    if(lfu->size >= lfu->capacity)
        removeLeastFrequent();

    Node* newnode = (Node*)malloc(sizeof(Node));
    strcpy(newnode->key, key);
    newnode->value = value;
    newnode->freq = 1;
    addToTail(newnode);
    updateFreq(key, value);
}

int get(int key) {
    if(lfu->size < lfu->capacity)
        return -1;

    Node* currNode = lfu->head;
    while(currNode != NULL && strcmp(currNode->key, key) != 0){
        currNode = currNode->next;
    }
    
    if(currNode == NULL || lfu->tail == NULL ||strcmp(currNode->key, key) != 0)
        return -1;

    removeNode(currNode);
    int freq = (currNode->freq + 1)%lfu->capacity;
    currNode->freq = freq;
    addToTail(currNode);

    return currNode->value;
}

void removeLeastFrequent(){
    Node* node = lfu->head;
    if(node == NULL) {
        free(node);
        return;
    }
    int minFreq = getCount(0);
    int firstKey = strcmp(node->key, lfu->head->key);
    if(firstKey < 0)
        minFreq++;
    
    while(lfu->tail != node && getCount(minFreq) > 1){
        removeNode(lfu->tail);
        lfu->size--;
        minFreq = (minFreq + 1)%lfu->capacity;
    }
    free(node);

}

void printLFU(LFUCache* lfu){
    Node* currNode = lfu->head;

    while(currNode != NULL) {
        printf("(%s, %d), ",currNode->key, currNode->value);
        if(getCount(0) > 1)
            printf("(");
        else
            printf(")");
        printf("(");
        for(int i=1; i<lfu->capacity; i++){
            if(getCount(i))
                printf("%d", i);
            printf(", ");
        }
        if(lfu->tail == currNode)
            printf(")\n");
        else
            printf("\n");
        currNode = currNode->next;
    }
}
```